sdk/core/Azure.Core/src/IJsonSerializable.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+
+namespace Azure
+{
+    /// <summary>
+    /// TODO
+    /// </summary>
+    public interface IJsonSerializable
+    {
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <param name="stream"></param>
+        /// <param name="bytesWritten"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        bool TrySerialize(Stream stream, out long bytesWritten, SerializableOptions? options = default);
```

Comment:
Did we consider also taking `BinaryData`?

Reply:
Could that handle a stream where position != 0 though? A `Utf8JsonWriter` could and supersedes a simple `PrettyPrint` in terms of flexibility.

Reply:
See https://github.com/Azure/azure-sdk-for-net/pull/35742#discussion_r1176840779

Reply:
In the similar vein - DynamicJson

Reply:
@pallavit - yes, it would be good to look at the interactions between DynamicData and the serialization interfaces and see if there's a good way to make it easy to go back and forth between them.  @m-nash and I have discussed this a tiny bit - it potentially lights up scenarios around going back and forth between protocol and convenience methods, and could also have perf implications where model types are large and callers only want to access a few properties on them when it doesn't make sense to deserialize the entire model.  For perf, I would expect us to want to use MutableJsonDocument rather than DynamicData, until there is lighter weight dynamic story.

In the meantime, DynamicData/MutableJsonDocument can serialize to a Stream, so this interface doesn't prevent that interaction, just adds a hop.

Does the motivation for taking `Stream` come from the fact that it is the fundamental vehicle for content (e.g. on `Response.ContentStream`?)

Reply:
The type we take in the public signature should not be coupled to any specific underlying serialization technology.  This is why something very generic like Stream was used as the baseline.  I do think other overloads are probably warranted we need to think through which ones we want to support with the only criteria being they don't couple us to any underlying serialization technology.

I mentioned in the description `string` is an interesting option, `BinaryData` could also be an options.

@heaths I do think we will need to validate the stream is seekable on input, @nisha-bhatia can you make a note for this update specifically.

Reply:
How did we land on using Try here? This doesn't match other popular serialization abstractions, such as STJ, or our ObjectSerializer abstraction.

Reply:
Why would it need to be seekable? I was merely stating that if `BinaryData` is meant to wrap an entire stream/writer/whatever, it may not work if we want to support serializing to a stream or writer where it's already not at the start e.g.,

```c#
using StreamWriter writer = new();
writer.Write(```{"foo":"bar", "baz":```);
azureSerializable.Serialize(writer.BaseStream);
writer.Write(```}```);
```

Reply:
+1 on why does the stream need to be seekable.
If the stream is not at position 0, wouldn't we just start writing at whatever the current position is?

Reply:
In the scenario of serialization the customer would want to read from the beginning after we have serialized into the stream if its not seekable how would they do this?

[Example usage](https://github.com/Azure/azure-sdk-for-net/pull/35742/files#diff-46013d0356c5137faeddc1f0b2c084532794fcea22ff8574fa4a46b824640855R65-R67)

Reply:
The stream might be a non-seekable FileStream that writes to disk.

Reply:
On the DynamicData pivot, is there a goal to keep the serialized-out format close enough to wire format that DynamicData could feasibly implement this interface?  Which is related to the question - should protocol method-only APIs have serializable "models" as well?